datafusion/functions-aggregate/src/nth_value.rs

```diff
@@ -87,27 +85,37 @@ impl AggregateUDFImpl for NthValueAgg {
     }
 
     fn accumulator(&self, acc_args: AccumulatorArgs) -> Result<Box<dyn Accumulator>> {
-        let n = match acc_args.input_exprs[1] {
-            Expr::Literal(ScalarValue::Int64(Some(value))) => {
-                if acc_args.is_reversed {
-                    Ok(-value)
-                } else {
-                    Ok(value)
+        let n = match acc_args.physical_exprs[1]
+            .as_any()
+            .downcast_ref::<Literal>()
+        {
+            Some(lit) => match lit.value() {
+                ScalarValue::Int64(Some(value)) => {
+                    if acc_args.is_reversed {
+                        -*value
+                    } else {
+                        *value
+                    }
+                }
+                _ => {
+                    return not_impl_err!(
+                        "{} not supported for n: {}",
+                        self.name(),
+                        &acc_args.physical_exprs[1]
+                    );
                 }
+            },
+            None => {
+                return not_impl_err!(
+                    "{} not supported for n: {}",
+                    self.name(),
+                    &acc_args.physical_exprs[1]
+                );
```

Comment:
The existing code is readable in the existing form. 

But I wanted to explore ways for removing the repetition in the error handling block. But ended up having to split it into two separate bindings to be able to reach there.

I don't have any strong opinions about this. So please feel free to use it or not :smiley: 

```rust
let scalar = acc_args.physical_exprs[1]
	.as_any()
	.downcast_ref::<Literal>()
	.and_then(|lit| {
		if let ScalarValue::Int64(inner) = lit.value() {
			Some(*inner)
		} else {
			None
		}
	})
	.flatten();

let n = if let Some(value) = scalar {
	value
} else {
	return not_impl_err!(
		"{} not supported for n: {}",
		self.name(),
		&acc_args.physical_exprs[1]
	);
};
```

Reply:
I come out another way that is slightly more readable